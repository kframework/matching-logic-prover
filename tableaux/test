#!/usr/bin/env bash

set -e

run_maude()  {
  maude </dev/null -no-banner -no-wrap <(echo 'set show stats off . set show timing off .') "$@"
}

do_diff() {
  git --no-pager diff --no-index --word-diff "$@"
}

run_test() {
  for t in "$@"; do
    test_actual=$(mktemp -t maude-run-XXXXXX)
    echo -n "$t ... "
    run_maude "$t" &> "$test_actual"
    echo "passed."
    do_diff "$test_actual" "$t.expected"
  done
  echo 'Passed.'
}

update_expected() {
  for t in "$@"; do
    run_maude "$t" &> "$t.expected"
  done
}

print_usage() {
    echo """
    Usage:

        $0
            run tests

        $0 update test_file1 [test_file2...]
            update expected output
    """
}

  if [[ "$#" = 0        ]] ; then run_test t/*.maude ;
elif [[ "$1" = "run"    ]] ; then shift; run_maude "$@" ;
elif [[ "$1" = "test"   ]] ; then shift; run_test "$@" ;
elif [[ "$1" = "update" ]] ; then shift; update_expected "$@" ;
elif [[ "$1" = "help"   ]] ; then shift; print_usage     "$@" ;
else print_usage "$@" ; echo >&2 "Unknown argument: $1"; exit 1;
fi
