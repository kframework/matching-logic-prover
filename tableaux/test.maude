--- Pattern ---
parse \not a .
parse \not \bottom .
parse \not \top .
parse < a > a .
parse < a > X .
parse [ a ] a .
parse [ a ] X .
parse \and(X, \and(Y, Z))  .
parse \and(\and(Y, Z), Z)  .
reduce \and(\and(Y, Z), Z)  .

--- SetVariable ---
reduce X in (X,Y) .
reduce X in (Y,Y) .

--- Free Variables ---
reduce fv(\and(X, Y)) .
reduce fv(\or(X, Y)) .
reduce fv(\nu X . X) .
reduce fv(\nu X . Y) .
reduce fv(\nu X . \and(X, Y)) .
reduce fv(\mu X . X) .
reduce fv(\mu X . Y) .
reduce fv(\mu X . \and(X, Y)) .
