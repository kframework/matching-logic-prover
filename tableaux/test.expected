Pattern: \not a
Pattern: < a > a
Pattern: < a > X{0}
Pattern: [a]a
Pattern: [a]X{0}
Pattern: \and(X{0}, \and(Y{0}, Z{0}))
Pattern: \and(\and(Y{0}, Z{0}), Z{0})
==========================================
reduce in MATCHING-LOGIC-PATTERN : \and(Z{0}, \and(Y{0}, Z{0})) .
rewrites: 0
result Pattern: \and(Y{0}, Z{0}, Z{0})
[Pattern]: \not \bottom
[Pattern]: \not \top
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] a .
rewrites: 1
result Symbol: a
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] Z{0} .
rewrites: 2
result IndexedSetVariable: Z{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] X{0} .
rewrites: 1
result IndexedSetVariable: Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] < a > X{0} .
rewrites: 3
result Pattern: < a > Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] < a > Y{0} .
rewrites: 4
result Pattern: < a > Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] < a > Z{0} .
rewrites: 4
result Pattern: < a > Z{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] [a]X{0} .
rewrites: 3
result Pattern: [a]Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] [a]Y{0} .
rewrites: 4
result Pattern: [a]Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] [a]Z{0} .
rewrites: 4
result Pattern: [a]Z{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \and(X{0}, Y{0}) .
rewrites: 7
result Pattern: \and(Y{0}, Y{0})
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \and(X{0}, Y{0}) .
rewrites: 7
result Pattern: \and(Y{0}, Y{0})
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \and(Y{0}, Z{0}) .
rewrites: 8
result Pattern: \and(Y{0}, Z{0})
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \or(X{0}, Y{0}) .
rewrites: 10
result Pattern: \or(Y{0}, Y{0})
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \or(X{0}, Y{0}) .
rewrites: 10
result Pattern: \or(Y{0}, Y{0})
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \or(Y{0}, Z{0}) .
rewrites: 11
result Pattern: \or(Y{0}, Z{0})
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \mu X . X{0} .
rewrites: 15
result Pattern: \mu * . *{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \mu X . Y{0} .
rewrites: 21
result Pattern: \mu * . Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \mu Z . X{0} .
rewrites: 20
result Pattern: \mu * . Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \mu Z . Y{0} .
rewrites: 21
result Pattern: \mu * . Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \mu Z . Z{0} .
rewrites: 15
result Pattern: \mu * . *{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \nu X . X{0} .
rewrites: 15
result Pattern: \nu * . *{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \nu X . Y{0} .
rewrites: 21
result Pattern: \nu * . Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \nu Z . X{0} .
rewrites: 20
result Pattern: \nu * . Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \nu Z . Y{0} .
rewrites: 21
result Pattern: \nu * . Y{0}
==========================================
reduce in MATCHING-LOGIC-SUBSTITUTION : [X := Y{0}] \nu Z . Z{0} .
rewrites: 15
result Pattern: \nu * . *{0}
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(X{0}) .
rewrites: 2
result DefinitionList: nil
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(a) .
rewrites: 2
result DefinitionList: nil
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(\and(X{0}, Y{0})) .
rewrites: 12
result DefinitionList: nil
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(\and(\mu X . \or(b, < a > X{0}), \mu Y . \or(b, < a > Y{0}))) .
rewrites: 140
result DefinitionConstant: u(0) := \mu * . \or(b, < a > *{0})
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(\and(\mu X . \or(b, < a > X{0}), \mu Y . \or(b, < c > Y{0}))) .
rewrites: 140
result NeDefinitionList: (u(0) := \mu * . \or(b, < a > *{0})) (u(1) := \mu * . \or(b, < c > *{0}))
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(\and(\mu X . \or(b, < a > X{0}), \mu X . \or(b, < a > X{0}))) .
rewrites: 114
result DefinitionConstant: u(0) := \mu * . \or(b, < a > *{0})
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(\mu X . \or(b, \nu Y . < a > Y{0})) .
rewrites: 111
result NeDefinitionList: (u(0) := \mu * . \or(b, \nu * . < a > *{0})) (u(1) := \nu * . < a > *{0})
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(\nu Y . < a > Y{0}) .
rewrites: 27
result DefinitionConstant: u(0) := \nu * . < a > *{0}
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : contract(\and(\mu X . \or([a]X{0}, \nu Y . \and(c, [a]Y{0})), \nu X . \and(< a > X{0}, \mu Y . \or(\not c, < a > Y{0})))) .
rewrites: 349
result NeDefinitionList: (u(0) := \mu * . \or([a]*{0}, \nu * . \and(c, [a]*{0}))) (u(1) := \nu * . \and(c, [a]*{0})) (u(2) := \nu * . \and(< a > *{0}, \mu * . \or(\not c, < a > *{0}))) (u(3) := \mu * . \or(\not c, < a > *{0}))
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : isConsistent(\and(\mu X . \or([a]X{0}, \nu Y . \and(c, [a]Y{0})), \nu X . \and(< a > X{0}, \mu Y . \or(\not c, < a > Y{0})))) .
rewrites: 164
result Bool: true
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : isConsistent(c, a, \not c) .
rewrites: 1
result Bool: false
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : isConsistent(a, c) .
rewrites: 9
result Bool: true
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : isConsistent(a, \not c) .
rewrites: 9
result Bool: true
==========================================
reduce in MATCHING-LOGIC-TABLEAUX : isReduced(\not c, < a > u(2){0}, [a]u(0){0}) .
rewrites: 40
result Bool: true
Bye.
